addons/mail/models/models.py

```diff
@@ -246,3 +247,108 @@ def _notify_email_header_dict(self):
+        """ Return cleaned create vals_list to avoid missing reference, creation of new record, deletion and update by:
+
+        - Removing set to any non-existing id for relations many2one and many2many if is_remove_missing_ref
+        - Removing set to non-existing fields
+        - Removing set to one2many relation (which will update the other side of the relation)
+        - Removing set that create sub-record
+        - Removing any command != LINK or SET (i.e. skipping CREATE, UPDATE, DELETE, UNLINK, CLEAR)
+
+        :param list|dict vals_list: value to be cleaned
+        :param bool is_remove_missing_ref: whether to remove missing reference or not (see above). If set to False, no
+        DB query is done and the method is fast.
+        """
+        if isinstance(vals_list, dict):
+            vals_list = [vals_list]
+
+        errors = defaultdict(list)
+
+        def filter_existing_ids(f_name, ids):
+            """ Filter existing ids if is_remove_missing_ref, otherwise return the ids as is.
+
+             :param str f_name: field name of this record
+             :param list ids: id of the target of the field relation
+             :return: If is_remove_missing_ref, returns only existing ids of the relation defined by f_name
+             (being active or not) otherwise returns the ids as is.
+             Note: Null (False, None, ...) are considered as existing ids.
+             """
+            if not is_remove_missing_ref:
+                return ids
+
+            def not_null(c_id):
+                """Check if the id is not null in the DB senses"""
+                return c_id is not False and c_id is not None  # touchy because False is an int and False == 0!
+
+            existing_ids_set = set(self.env[self._fields[f_name].comodel_name]
+                                   .browse(filter(not_null, ids)).exists().mapped('id'))
+            filtered_existing_ids = [c_id for c_id in ids if c_id in existing_ids_set or not not_null(c_id)]
+            if filtered_existing_ids != ids:
+                missing_references = set(ids) - set(filtered_existing_ids)
+                errors[f_name].append(
+                    ('MISSING_REF', _('missing reference (%s)', ','.join(str(r) for r in missing_references))))
+            return filtered_existing_ids
+
+        new_vals_list = []
+        for vals in vals_list:
+            vals_filtered = dict()
+            for field_name, value in vals.items():
+                new_value = None
+                field = self._fields.get(field_name)
+                if not field:
+                    errors[field_name].append(('MISSING_FIELD', _('missing field')))
+                    continue
+                if field.type == 'one2many':
+                    errors[field_name].append(('NOT_ALLOWED_ONE2MANY', _('One2Many field not allowed')))
+                    continue
+
+                # Inspired from fields.convert_to_cache
```

Comment:
Why don't you try to actually use `field.convert_to_cache()` ?
This might cover 90% of the issues.
Discarding existing ids for relational fields will provide some extra cleanup.